doc/media-settings-enhancements/media-settings-plugin.md

```diff
@@ -0,0 +1,20 @@
+# Enhancement for Media Settings Lookup Key
+## Introduction
+The SONiC optical media transceiver daemon, xcvrd, is responsible for obtaining certain settings that are dependent on the type of transceiver installed, which port it is installed in, and on which platform it is running.  An example of such a setting is NPU port preemphasis.  
+This is done using a JSON file, media_settings.json, provided with each platform, to describe a mapping from certain transceiver properties to media settings.  Xcvrd builds pair of keys, based on the type of the inserted transceiver, and for each key, attempts to find an entry in a dictionary created from media_settings.json for that key; if an entry is found, the settings corresponding to that key are used.
+## Current implementation
+In xcvrd, the function get_media_settings_key() is used to form a list of two strings.  These strings are used as lookup keys for the dictionary created from the file media_settings.json.  
+The current implementation of get_media_settings_key() constructs two keys, as follows:
+- the first key is of the form _manufacturer_-_partnumber_, where _manufacturer_ and _partnumber_ are the name of the transceiver manufacturer and the manufacturer's part number, respectively, which are read from the transceiver EEPROM, and
+- the second key if of the form _type_-_compliancecode_-_length_, where _type_ is the transceiver type (e.g. SFP, SFP28, QSFP, etc.), _compliancecode_ is the physical layer standard (e.g. 10GBASE-SR, 40GBASE-CR4), and _length_ is the length of the attached cable, if applicable.
+# Enhancement
+Xcvrd shall be modified such that it will check for the presence of a platform-dependent plugin, which shall contain an implementation of the function get_media_settings_key().  Xcvrd shall check for the presence of this plugin, and, if found, use it in preference to the built-in function get_media_settings_key().  
```

Comment:
what is the motivation to introduce the platform-dependent plugin. it is unclear to me.

Reply:
The current code in xcvrd to form the media settings lookup keys does not work -- for example, for a 1 m SFP28 DAC cable, it returns a key of "SFP28", no indication that it is a copper cable, no information about the the cable length, and both copper/fiber information and cable length information are needed to make a preemphasis lookup.
In order to maintain backward compatibility, i.e. still be able to process media_settings.json in the current way for existing platforms, an optional plugin for xcvrd can be added, such that platforms that cannot work with the existing key-formation logic may properly look up their media settings.
The alternative is to have in-line code in xcvrd that both has the the existing "broken" logic, and new logic, to properly form lookup keys.

Reply:
this looks like a generic improvement, it should not be platform dependent. can you describe the motivation in the doc. also I suggest to move to the same folder as the previous design doc.

Reply:
prefer inline code.

Reply:
The real issue backward compatibility.

While in principle, there is no reason that mainline xcvrd code should not be able to construct a lookup key that will account for all possible media types in used in all platforms, the fact of the matter is that the current implementation of get_media_settings_key() is flawed, and media_settings.json files for existing platforms rely on the current method for building the key.  Therefore, in order to maintain compatibility with existing platforms, an lookup implemented in mainline code must (1) build a key in the old style, (2) attempt a lookup, (3) if a hit is not found, build a key in a newer style, and (4) attempt a lookup.  That newer style could be something similar to the existing style, but more complete, i.e. QSFP28-100GBASE-CR-1M for a 1 m 100G DAC cable.  Also, any entry "Default" entry in the media_settings.json file would not be used unless both the lookup for an old-style and new-style key failed.
This need to keep backward compatible was making xcvrd rather convoluted -- moving to a platform-dependent plugin seemed cleaner.  Also, media_settings.json is a platform-dependent file... so the fragment of code to generate the lookup key for it being platform-dependent is not wholly unreasonable...